File: PCVR/R/pcvpca.R
#' Procrustes cross-validation for PCA models
#'
#' @param X
#' matrix with predictors from the training set.
#' @param ncomp
#' number of components to use (more than the expected optimal number).
#' @param center
#' logical, to center or not the data sets
#' @param scale
#' logical, to scale or not the data sets
#' @param cv
#' which split method to use for cross-validation (see description for details).
#' @param cv.scope
#' scope for center/scale operations inside CV loop: 'global' — using globally computed mean and std
#' or 'local' — recompute new for each local calibration set.
#'
#' @details
#' The method computes pseudo-validation matrix Xpv, based on PCA decomposition of calibration
#' set `X` and cross-validation. See description of the method in [1].
#'
#' Parameter `cv` defines how to split the rows of the training set. The split is similar
#' to cross-validation splits, as PCV is based on cross-validation. This parameter can have
#' the following values:
#'
#' 1. A number (e.g. `cv = 4`). In this case this number specifies number of segments for random
#' splits, except `cv = 1` which is a special case for leave-one-out (full cross-validation).
#'
#' 2. A list with 2 values: `list("name", nseg)`. In this case `"name"` defines the way to make
#' the split, you can select one of the following: `"loo"` for leave-one-out, `"rand"` for random
#' splits or `"ven"` for Venetian blinds (systematic) splits. The second parameter, `nseg`, is a
#' number of segments for splitting the rows into. For example, `cv = list("ven", 4)`, shown in
#' the code examples above, tells PCV to use Venetian blinds splits with 4 segments.
#'
#' 3. A vector with integer numbers, e.g. `cv = c(1, 2, 3, 1, 2, 3, 1, 2, 3)`. In this case number
#' of values in this vector must be the same as number of rows in the training set. The values
#' specify which segment a particular row will belong to. In case of the example shown here, it
#' is assumed that you have 9 rows in the calibration set, which will be split into 3 segments.
#' The first segment will consist of measurements from rows 1, 4 and 7.
#'
#' Parameter `cv.scope` influences how the Procrustean rule is met. In case of "global" scope,
#' the rule will be met strictly - error of predictions for PV-set and the global model will be
#' identical to the error from conventional cross-validation. In case of "local" scope, every
#' local model will have its own center and scaling factor and hence the rule will be almost
#' met (the errors will be close but not identical).
#'
#' @return
#' Matrix with PV-set (same size as X)
#'
#' @references
#' 1. S. Kucheryavskiy, O. Rodionova, A. Pomerantsev. Procrustes cross-validation of multivariate
#' regression models. Submitted, 2022.
#'
#' @examples
#'
#' # load NIR spectra of Corn samples
#' data(corn)
#' X <- corn$spectra
#'
#' # generate Xpv set based on PCA decomposition with A = 20 and venetian blinds split with 4 segments
#' Xpv <- pcvpca(X, ncomp = 20, center = TRUE, scale = FALSE, cv = list("ven", 4))
#'
#' # show the original spectra and the PV-set (as is and mean centered)
#' oldpar <- par(mfrow = c(2, 2))
#' matplot(t(X), type = "l", lty = 1, main = "Original data")
#' matplot(t(Xpv), type = "l", lty = 1, main = "PV-set")
#' matplot(t(scale(X, scale = FALSE)), type = "l", lty = 1, main = "Original data (mean centered)")
#' matplot(t(scale(Xpv, scale = FALSE)), type = "l", lty = 1, main = "PV-set (mean centered)")
#' par(oldpar)
#'
#' @importFrom stats sd
#'
#' @export
pcvpca <- function(X, ncomp = min(nrow(X) - 1, ncol(X), 30), cv = list("ven", 4),
   center = TRUE, scale = FALSE, cv.scope = "global") {

   # keep names if any
   attrs <- attributes(X)

   # compute global mean and standard deviation and autoscale the whole data
   mX <- if (center) apply(X, 2, mean) else rep(0, ncol(X))
   sX <- if (scale) apply(X, 2, sd) else rep(1, ncol(X))
   X <- scale(X, center = mX, scale = sX)

   # indices for cross-validation
   ind <- pcvcrossval(cv, nrow(X), seq_len(nrow(X)))

   # get number of segments
   nseg <- max(ind)

   # correct maximum number of components
   max.ncomp <- min(nrow(X) - ceiling(nrow(X) / nseg) - 1 , ncol(X))
   if (ncomp > max.ncomp) {
      ncomp <- max.ncomp
   }

   # create a global model
   m <- svd(X, nv = ncomp, nu = ncomp)
   s <- m$d[seq_len(ncomp)]
   P <- m$v
   PPT <- tcrossprod(P)
   Pi <- diag(1, nrow(P)) - PPT

   # prepare empty matrix for pseudo-validation set
   Xpv <- matrix(0, nrow(X), ncol(X))

   # loop for computing the PV set
   for (k in seq_len(nseg)) {

      # split data to calibration and validation
      ind.c <- ind != k
      ind.k <- ind == k

      X.c <- X[ ind.c, , drop = FALSE]
      X.k <- X[ ind.k, , drop = FALSE]

      # compute mean and standard deviation and autoscale in case of local scope
      if (cv.scope == "local") {
         mXl <- if (center) apply(X.c, 2, mean) else rep(0, ncol(X.c))
         sXl <- if (scale) apply(X.c, 2, sd) else rep(1, ncol(X.c))
         X.c <- scale(X.c, center = mXl, scale = sXl)
         X.k <- scale(X.k, center = mXl, scale = sXl)
      }

      # get loadings for local model and rotation matrix between global and local models
      m.k <- svd(X.c, nv = ncomp, nu = ncomp)
      P.k <- m.k$v

      # correct direction of loadings for local model
      a <- acos(colSums(P * P.k)) < pi / 2
      P.k <- P.k %*% diag(a * 2 - 1, ncol(P), ncol(P))

      # get scores and residuals by projection local validation set to the local model
      T.k <- X.k %*% P.k
      E.k <- X.k - tcrossprod(T.k, P.k)
      q.k <- rowSums(E.k^2)

      # compute the parallel component of Xpv
      Xpv.hat <- tcrossprod(T.k, P)

      # compute the orthogonal component of Xpv
      Xpv.orth <- getxpvorth(q.k, X.k, Pi)

      # create and save the Xpv
      Xpv[ind.k, ] <- Xpv.hat + Xpv.orth
   }

   # uscenter and unscale the data using global mean and std
   Xpv <- sweep(Xpv, 2, sX, "*")
   Xpv <- sweep(Xpv, 2, mX, "+")

   attributes(Xpv) <- attrs
   return(Xpv)
}

File: PCVMATLAB/pcvpca.m
function Xpv = pcvpca(X, nComp, Center, Scale, CV, CVScope)
%PCVPCA Compute matrix with pseudo-validation set for PCA/SIMCA model
%
% Arguments:
% ----------
% X         matrix with calibration set
% nComp     number of components for PCA decomposition
% Center    logical, mean center data columns or not
% Scale     logical, standardize data columns or not
% CV        defines cross-validation splits pattern (see description)
% CVScope   defines if centering/scaling must be done globally ('global')
% or locally ('local'). Default value is 'global'.
%
% Returns:
% --------
% Xpv       matrix with PV-set (same size as X)
%
% The method computes pseudo-validation matrix Xpv, based on PCA decomposition of
% calibration set X and cross-validation. The 'CV' parameter can be:
%
% A number:
% =========
% 'CV = 1' - full cross-validation (leave one out)
% 'CV = n' - random cross-validation with n segments
%
% A cell array:
% =============
% 'CV = {"rand", n}' - random cross-validation with n segments
% 'CV = {"ven", n}' - systematic cross-validation with n segments
% 'CV = {"loo"}' - full cross-validation (leave one out)
%
% A vector:
% =========
% 'CV = [1, 2, 3, 1, 2, 3] - cross-validation of 6 rows with 3 segments
%
% See more details here: https://github.com/svkucheryavski/pcv
%
% See also: PCVPCR, PCVPLS

   nRows = size(X, 1);
   nCols = size(X, 2);

   % default values
   if nargin < 6
      CVScope = "global";
   end
   
   if nargin < 5
       CV = {"ven", 4};
   end

   if nargin < 4
       Scale = false;
   end

   if nargin < 3
       Center = true;
   end

   % compute indices for cross-validation
   ind = crossval(CV, nRows);

   % compute number of segments
   nSeg = max(ind);

   maxNComp = min([nRows - round(nRows / nSeg) - 1, nCols, 30]);

   if nargin < 2 || nComp > maxNComp
      nComp = maxNComp;
   end

   % compute and save mean and standard deviation
   if Center
      mX = mean(X);
   else
      mX = zeros(1, nCols);
   end

   if Scale
      sX = std(X);
   else
      sX = ones(1, nCols);
   end

   % autoscale data globally
   X = (X - mX) ./ sX;
   
   % create a global model
   [~, ~, P] = svds(X, nComp);
   Pi = eye(nCols) - P * P';

   % prepare empty matrix for pseudo-validation set
   Xpv = zeros(nRows, nCols);

   % cv-loop
   for k = 1:nSeg

      % split data to calibration and validation
      indc = ind ~= k;
      indk = ind == k;

      Xk = X(indk, :);
      Xc = X(indc, :);
      
      % compute mean and standard deviation and autoscale locally in case of local scope
      if CVScope == "local"
         if Center, mXl = mean(Xc); else, mXl = zeros(1, nCols); end
         if Scale, sXl = std(Xc); else, sXl = ones(1, nCols); end
         Xc = (Xc - mXl) ./ sXl;
         Xk = (Xk - mXl) ./ sXl;
      end
      
      % get loadings for local model and rotation matrix between global and local models
      [~, ~, Pk] = svds(Xc, nComp);

      % correct direction of loadings for local model
      a = acos(sum(P .* Pk)) < pi / 2;
      Pk = Pk * diag(a * 2 - 1);

      % compute explained part of Xpv
      Tk = Xk * Pk;
      Xpvpar = Tk * P';
            
      % compute orthogonal part of Xpv
      Ek = Xk - Tk * Pk';
      qk = sum(Ek.^2, 2);
      Xpvorth = getxpvorth(qk, Xk, Pi);

      % rotate the local validation set and save as a part of Xpv
      Xpv(indk, :) = Xpvorth + Xpvpar;      
   end

   % uscenter and unscale the data
   Xpv = Xpv .* sX + mX;
end

